File: trunk/Gomoku/Client/cpp/Gomoku/AI/Source/JamesCook.cpp
/*
 *  Gomoku Core
 * 
 *  Copyright (c) 2011 Tran Dinh Thoai <dthoai@yahoo.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include "..\Header\JamesCook.h"

using namespace gomoku_ai;

const int JamesCook::WIN_COUNT = 5;

JamesCook::JamesCook(Config *config, Board *board, unsigned char piece) : Player(config, board, piece) {

}

JamesCook::~JamesCook() {

}

void JamesCook::think(Move &move) {
	if (!this->getReady()) return;
	Cell *cell = this->findBestMoves(*this->board);
	if (cell != NULL) {
		Move *bestMove = new Move(cell->getRow(), cell->getColumn(), this->getPiece());
		move.clone(*bestMove);
		this->makeMove(move);
	}
}

void JamesCook::getLines() {
	if (this->lines.size() > 0) return;

	int i, j, n;

	for (i = 0; i < this->board->getHeight(); i++) {
		Line *line = new Line();
		for (j = 0; j < this->board->getWidth(); j++) {
			line->add(i, j);
		}
		this->lines.add(line);
	}
	for (j = 0; j < this->board->getWidth(); j++) {
		Line *line = new Line();
		for (i = 0; i < this->board->getHeight(); i++) {
			line->add(i, j);
		}
		this->lines.add(line);
	}
	for (n = 0; n < this->board->getHeight(); n++) {
		i = n;
		j = 0;
		Line *line = new Line();
		while (i < this->board->getHeight() && j < this->board->getWidth()) {
			line->add(i, j);
			i++;
			j++;
		}
		this->lines.add(line);
	}
	for (n = 1; n < this->board->getWidth(); n++) {
		i = 0;
		j = n;
		Line *line = new Line();
		while (i < this->board->getHeight() && j < this->board->getWidth()) {
			line->add(i, j);
			i++;
			j++;
		}
		this->lines.add(line);
	}
	for (n = 0; n < this->board->getHeight(); n++) {
		i = n;
		j = this->board->getWidth() - 1;
		Line *line = new Line();
		while (i < this->board->getHeight() && j >= 0) {
			line->add(i, j);
			i++;
			j--;
		}
		this->lines.add(line);
	}
	for (n = this->board->getWidth() - 2; n >= 0; n--) {
		i = 0;
		j = n;
		Line *line = new Line();
		while (i < this->board->getHeight() && j >= 0) {
			line->add(i, j);
			i++;
			j--;
		}
		this->lines.add(line);
	}
}

JamesCook::Row *JamesCook::getBlanks() {
	Row *tag = new Row();
		
	for (int i = 0; i < this->board->getHeight(); i++) {
		for (int j = 0; j < this->board->getWidth(); j++) {
			if (this->board->getPiece(i, j) == Board.BLANK) {
				tag->Cells.push_back(new Cell(i, j));
			}
		}
	}
		
	return tag;
}

Cell *JamesCook::findBestMoves(Board &board) {
	Row *computerMoves = new Row();
	Row *humanMoves = new Row();
	int computerCounter = this->findBestMoves(this->getPiece(), board, *computerMoves);
	int humanCounter = this->findBestMoves(Board::opponentPiece(this->getPiece()), board, *humanMoves);
		
	if (humanMoves->Cells.size() == 0) {
		if (computerMoves->Cells.size() > 0) {
			return computerMoves->getCell(0);
		} else {
			Row *blanks = this->getBlanks();
			if (blanks->Cells.size() == this->board->getWidth() * this->board->getHeight()) {
				int row = this->board->getHeight() / 2;
				int column = this->board->getWidth() / 2;
				return new Cell(row, column);
			} else if (blanks->Cells.size() > 0) {
					    return blanks->getCell(0);
			}
		}
	} else if (computerMoves->Cells.size() == 0) {
		return humanMoves->getCell(0);
	} else {
		if (humanCounter >= computerCounter) {
			return humanMoves->getCell(0);
		} else {
			return computerMoves->getCell(0);
		}
	}

	return NULL;
}

int JamesCook::findBestMoves(int srcState, Board &board, Row &tag) {
	int counter = 0;
	RowList rows;
	Row *row;
	int i;

	this->getLines();

	for (i = 0; i < this->lines.size(); i++) {
		Line *line = this->lines.get(i);
		getRows(srcState, board, *line, rows);
	}
		
	int maxCount = 0;
	for (i = rows.size() - 1; i >= 0; i--) {
		row = rows.get(i);
		if (row->BlankStart.size() == 0 && row->BlankStop.size() == 0) {
			rows.remove(i);
			continue;
		}
		if (row->Cells.size() > WIN_COUNT) {
			rows.remove(i);
			continue;
		}
		if (row->BlankStart.size() + row->BlankStop.size() + row->Cells.size() < WIN_COUNT) {
			rows.remove(i);
			continue;
		}
		if (maxCount < row->Cells.size()) {
			maxCount = row->Cells.size();
		}
	}
	counter = maxCount;
	for (i = rows.size() - 1; i >= 0; i--) {
		row = rows.get(i);
		if (maxCount > row->Cells.size()) {
			rows.remove(i);
			continue;
		}
	}
	for (i = 0; i < rows.size(); i++) {
		row = rows.get(i);
		if (row->BlankStart.size() == 0 || row->BlankStop.size() == 0) continue;
		if (row->BlankStart.size() > 0) {
			tag.Cells.push_back(row->getStart(row->BlankStart.size() - 1));
		}
		if (row->BlankStop.size() > 0) {
			tag.Cells.push_back(row->getStop(0));
		}
	}
	if (tag.Cells.size() == 0) {
		for (i = 0; i < rows.size(); i++) {
			row = rows.get(i);
			if (row->BlankStart.size() > 0 && row->BlankStop.size() > 0) continue;
			if (row->BlankStart.size() > 0) {
				tag.Cells.push_back(row->getStart(row->BlankStart.size() - 1));
			}
			if (row->BlankStop.size() > 0) {
				tag.Cells.push_back(row->getStop(0));
			}
		}
	}
		
	return counter;
}

void JamesCook::getRows(int srcState, Board &board, Line &line, RowList &tag) {
	Row *row = new Row();
	bool start = false;
	int i, j;
		
	for (i = 0; i < line.size(); i++) {
		int tagState = this->board->getPiece(line.get(i)->getRow(), line.get(i)->getColumn());
		if (start) {
			if (tagState == srcState) {
				row->Cells.push_back(line.get(i));
				if (i == line.size() - 1) {
					tag.add(row);
					row = new Row();
				}
			} else {
				for (j = i; j < line.size(); j++) {
					if (this->board->getPiece(line.get(j)->getRow(), line.get(j)->getColumn()) == Board::BLANK) {
						row->BlankStop.push_back(line.get(j));
					} else {
						break;
					}
				}
				tag.add(row);
				row = new Row();
				if (tagState == Board::BLANK) {
					row->BlankStart.push_back(line.get(i));
				}
				start = false;
			}
		} else {
			if (tagState == srcState) {
				start = true;
				row->Cells.push_back(line.get(i));
			} else if (tagState == Board::BLANK) {
				row->BlankStart.push_back(line.get(i));
			} else {
				row = new Row();
			}
		}
	}
}

Cell *JamesCook::Row::getCell(int idx) {
	list<Cell *>::iterator it;
	int no = 0;
	for (it = this->Cells.begin(); it != this->Cells.end(); it++) {
		if (no == idx) {
			return (*it);
		}
		no++;
	}
	return NULL;
}

Cell *JamesCook::Row::getStart(int idx) {
	list<Cell *>::iterator it;
	int no = 0;
	for (it = this->BlankStart.begin(); it != this->BlankStart.end(); it++) {
		if (no == idx) {
			return (*it);
		}
		no++;
	}
	return NULL;
}

Cell *JamesCook::Row::getStop(int idx) {
	list<Cell *>::iterator it;
	int no = 0;
	for (it = this->BlankStop.begin(); it != this->BlankStop.end(); it++) {
		if (no == idx) {
			return (*it);
		}
		no++;
	}
	return NULL;
}

int JamesCook::RowList::size() {
	return this->rows.size();
}

JamesCook::Row *JamesCook::RowList::get(int idx) {
	list<Row *>::iterator it;
	int no = 0;
	for (it = this->rows.begin(); it != this->rows.end(); it++) {
		if (no == idx) {
			return (*it);
		}
		no++;
	}
	return NULL;
}

void JamesCook::RowList::add(Row *row) {
	this->rows.push_back(row);
}

void JamesCook::RowList::remove(int idx) {
	list<Row *>::iterator it;
	int no = 0;
	for (it = this->rows.begin(); it != this->rows.end(); it++) {
		if (no == idx) {
			this->rows.remove(*it);
			return;
		}
		no++;
	}
}

int JamesCook::Line::size() {
	return this->cells.size();
}

Cell *JamesCook::Line::get(int idx) {
	list<Cell *>::iterator it;
	int no = 0;
	for (it = this->cells.begin(); it != this->cells.end(); it++) {
		if (no == idx) {
			return (*it);
		}
		no++;
	}
	return NULL;
}

void JamesCook::Line::add(int row, int column) {
	Cell *cell = new Cell(row, column);
	this->cells.push_back(cell);
}

int JamesCook::LineList::size() {
	return this->lines.size();
}

JamesCook::Line *JamesCook::LineList::get(int idx) {
	list<Line *>::iterator it;
	int no = 0;
	for (it = this->lines.begin(); it != this->lines.end(); it++) {
		if (no == idx) {
			return (*it);
		}
		no++;
	}
	return NULL;
}

void JamesCook::LineList::add(Line *line) {
	this->lines.push_back(line);
}

File: Gomoku/Client/vb/Gomoku/AI/JamesCook.vb
'
'  Gomoku Core
' 
'  Copyright (c) 2011 Tran Dinh Thoai <dthoai@yahoo.com>
'
' This library is free software; you can redistribute it and/or
' modify it under the terms of the GNU General Public License
' version 3.
'
' This library is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
' General Public License for more details.
'
' You should have received a copy of the GNU General Public License
' along with this library; if not, write to the Free Software
' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'

Option Explicit On
Option Strict On

Imports com.bhivef.gomoku.core

Namespace com.bhivef.gomoku.ai

    Public Class JamesCook
        Inherits Player

        Protected Const WIN_COUNT As Integer = 5
        Protected lines As List(Of Line)

        Public Sub New(ByVal config As Config, ByVal board As Board, ByVal piece As Byte)
            MyBase.New(config, board, piece)
        End Sub

        Public Overrides Sub think(ByVal move As Move)
            If Not getReady() Then
                Return
            End If
            Dim cell As Cell = findBestMoves(board)
            If cell IsNot Nothing Then
                makeMove(New Move(cell.getRow(), cell.getColumn(), getPiece()))
            End If
        End Sub

        Protected Function getLines() As List(Of Line)
            If lines IsNot Nothing Then
                Return lines
            End If

            lines = New List(Of Line)

            For i As Integer = 0 To board.getHeight() - 1
                Dim line As New Line()
                For j As Integer = 0 To board.getWidth() - 1
                    line.add(i, j)
                Next
                lines.Add(line)
            Next
            For j As Integer = 0 To board.getWidth() - 1
                Dim line As New Line()
                For i As Integer = 0 To board.getHeight() - 1
                    line.add(i, j)
                Next
                lines.Add(line)
            Next
            For n As Integer = 0 To board.getHeight() - 1
                Dim i As Integer = n
                Dim j As Integer = 0
                Dim line As New Line()
                While i < board.getHeight() And j < board.getWidth()
                    line.add(i, j)
                    i += 1
                    j += 1
                End While
                lines.Add(line)
            Next
            For n As Integer = 1 To board.getWidth() - 1
                Dim i As Integer = 0
                Dim j As Integer = n
                Dim line As New Line()
                While i < board.getHeight() And j < board.getWidth()
                    line.add(i, j)
                    i += 1
                    j += 1
                End While
                lines.Add(line)
            Next
            For n As Integer = 0 To board.getHeight() - 1
                Dim i As Integer = n
                Dim j As Integer = board.getWidth() - 1
                Dim line As New Line()
                While i < board.getHeight() And j >= 0
                    line.add(i, j)
                    i += 1
                    j -= 1
                End While
                lines.Add(line)
            Next
            For n As Integer = board.getWidth() - 2 To 0 Step -1
                Dim i As Integer = 0
                Dim j As Integer = n
                Dim line As New Line()
                While i < board.getHeight() And j >= 0
                    line.add(i, j)
                    i += 1
                    j -= 1
                End While
                lines.Add(line)
            Next
            Return lines
        End Function

        Protected Function getBlanks() As List(Of Cell)
            Dim tag As New List(Of Cell)
            For i As Integer = 0 To board.getHeight() - 1
                For j As Integer = 0 To board.getWidth() - 1
                    If board.getPiece(i, j) = core.Board.BLANK Then
                        tag.Add(New Cell(i, j))
                    End If
                Next
            Next
            Return tag
        End Function

        Protected Function findBestMoves(ByVal board As Board) As Cell
            Dim computerMoves As New List(Of Cell)
            Dim humanMoves As New List(Of Cell)
            Dim computerCounter As Integer = findBestMoves(getPiece(), board, computerMoves)
            Dim humanCounter As Integer = findBestMoves(core.Board.opponentPiece(getPiece()), board, humanMoves)

            If humanMoves.Count = 0 Then
                If computerMoves.Count > 0 Then
                    Return computerMoves(0).clone()
                Else
                    Dim blanks As List(Of Cell) = getBlanks()
                    If blanks.Count = board.getWidth() * board.getHeight() Then
                        Dim row As Integer = CType(board.getHeight() / 2, Integer)
                        Dim column As Integer = CType(board.getWidth() / 2, Integer)
                        Return New Cell(row, column)
                    ElseIf blanks.Count > 0 Then
                        Return blanks(0).clone()
                    End If
                End If
            ElseIf computerMoves.Count = 0 Then
                Return humanMoves(0).clone()
            Else
                If humanCounter >= computerCounter Then
                    Return humanMoves(0).clone()
                Else
                    Return computerMoves(0).clone()
                End If
            End If

            Return Nothing
        End Function

        Protected Function findBestMoves(ByVal srcState As Integer, ByVal board As Board, ByVal tag As List(Of Cell)) As Integer
            Dim counter As Integer = 0
            Dim rows As New List(Of Row)
            Dim lines As List(Of Line) = getLines()

            For i As Integer = 0 To lines.Count - 1
                getRows(srcState, board, lines(i), rows)
            Next

            Dim maxCount As Integer = 0
            For i As Integer = rows.Count - 1 To 0 Step -1
                Dim row As Row = rows(i)
                If row.BlankStart.Count = 0 And row.BlankStop.Count = 0 Then
                    rows.RemoveAt(i)
                    Continue For
                End If
                If row.Cells.Count > WIN_COUNT Then
                    rows.RemoveAt(i)
                    Continue For
                End If
                If row.BlankStart.Count + row.BlankStop.Count + row.Cells.Count < WIN_COUNT Then
                    rows.RemoveAt(i)
                    Continue For
                End If
                If maxCount < row.Cells.Count Then
                    maxCount = row.Cells.Count
                End If
            Next
            counter = maxCount
            For i As Integer = rows.Count - 1 To 0 Step -1
                Dim row As Row = rows(i)
                If maxCount > row.Cells.Count Then
                    rows.RemoveAt(i)
                    Continue For
                End If
            Next
            For i As Integer = 0 To rows.Count - 1
                Dim row As Row = rows(i)
                If row.BlankStart.Count = 0 Or row.BlankStop.Count = 0 Then
                    Continue For
                End If
                If row.BlankStart.Count > 0 Then
                    tag.Add(row.BlankStart(row.BlankStart.Count - 1).clone())
                End If
                If row.BlankStop.Count > 0 Then
                    tag.Add(row.BlankStop(0).clone())
                End If
            Next
            If tag.Count = 0 Then
                For i As Integer = 0 To rows.Count - 1
                    Dim row As Row = rows(i)
                    If row.BlankStart.Count > 0 And row.BlankStop.Count > 0 Then
                        Continue For
                    End If
                    If row.BlankStart.Count > 0 Then
                        tag.Add(row.BlankStart(row.BlankStart.Count - 1).clone())
                    End If
                    If row.BlankStop.Count > 0 Then
                        tag.Add(row.BlankStop(0).clone())
                    End If
                Next
            End If

            Return counter
        End Function

        Protected Sub getRows(ByVal srcState As Integer, ByVal board As Board, ByVal line As Line, ByVal tag As List(Of Row))
            Dim row As New Row()
            Dim start As Boolean = False

            For i As Integer = 0 To line.size() - 1
                Dim tagState As Integer = board.getPiece(line.get(i).getRow(), line.get(i).getColumn())
                If start Then
                    If tagState = srcState Then
                        row.Cells.Add(line.get(i).clone())
                        If i = line.size() - 1 Then
                            tag.Add(row)
                            row = New Row()
                        End If
                    Else
                        For j As Integer = i To line.size() - 1
                            If board.getPiece(line.get(j).getRow(), line.get(j).getColumn()) = core.Board.BLANK Then
                                row.BlankStop.Add(line.get(j).clone())
                            Else
                                Exit For
                            End If
                        Next
                        tag.Add(row)
                        row = New Row()
                        If tagState = core.Board.BLANK Then
                            row.BlankStart.Add(line.get(i).clone())
                        End If
                        start = False
                    End If
                Else
                    If tagState = srcState Then
                        start = True
                        row.Cells.Add(line.get(i).clone())
                    ElseIf tagState = core.Board.BLANK Then
                        row.BlankStart.Add(line.get(i).clone())
                    Else
                        row = New Row()
                    End If
                End If
            Next
        End Sub

        Protected Class Row
            Public Cells As New List(Of Cell)
            Public BlankStart As New List(Of Cell)
            Public BlankStop As New List(Of Cell)

            Public Overrides Function ToString() As String
                Dim tag As String = ""
                tag &= "\r\nCells: "
                For i As Integer = 0 To Cells.Count - 1
                    tag &= Cells(i).ToString() & " "
                Next
                tag &= "\r\nBlankStart: "
                For i As Integer = 0 To BlankStart.Count - 1
                    tag &= BlankStart(i).ToString() & " "
                Next
                tag &= "\r\nBlankStop: "
                For i As Integer = 0 To BlankStop.Count - 1
                    tag &= BlankStop(i).ToString() & " "
                Next
                Return tag
            End Function
        End Class

        Protected Class Line
            Private cells As List(Of Cell)

            Public Sub New()
                cells = New List(Of Cell)
            End Sub

            Public Function size() As Integer
                Return cells.Count
            End Function

            Public Function [get](ByVal idx As Integer) As Cell
                Return cells(idx).clone()
            End Function

            Public Sub add(ByVal row As Integer, ByVal column As Integer)
                cells.Add(New Cell(row, column))
            End Sub

            Public Sub add(ByVal src As Cell)
                cells.Add(src.clone())
            End Sub

            Public Sub remove(ByVal idx As Integer)
                cells.RemoveAt(idx)
            End Sub

            Public Function indexOf(ByVal src As Cell) As Integer
                For i As Integer = 0 To cells.Count - 1
                    If cells(i).equals(src) Then
                        Return i
                    End If
                Next
                Return -1
            End Function

            Public Function indexOf(ByVal row As Integer, ByVal column As Integer) As Integer
                Return indexOf(New Cell(row, column))
            End Function
        End Class

    End Class

End Namespace